test/compiler/inference.jl

```diff
@@ -3393,6 +3393,98 @@ end
     end) == Any[Int]
 end
 
+Base.@noinfer count_specialization(@nospecialize f) = count_specialization(method1(f))
+count_specialization(m::Method) = count(!isnothing, m.specializations)
+Base.@noinfer method1(@nospecialize f) = only(methods(f))
+Base.@noinfer code_typed1(@nospecialize args...) = first(only(code_typed(args...))).code
+using InteractiveUtils: gen_call_with_extracted_types_and_kwargs
+macro code_typed1(ex0...)
+    gen_call_with_extracted_types_and_kwargs(__module__, :code_typed1, ex0)
+end
+
+@testset "compilation annotations" begin
+    m = Module()
+
+    dispatchonly = Any[sin, muladd, "foo", nothing, Dict]   # untyped container can cause excessive runtime dispatch
+    withinfernce = tuple(sin, muladd, "foo", nothing, Dict) # typed container can cause excessive inference
+
+    @eval m begin
+        function invokef(f, itr)
+            local r = 0
+            r += f(itr[1])
+            r += f(itr[2])
+            r += f(itr[3])
+            r += f(itr[4])
+            r += f(itr[5])
+            r
+        end
+        global inline_checker = c -> c # untyped global prevents inlining
+        # if `f` is inlined, `GlobalRef(m, :inline_checker)` should appear within the body of `invokef`
+        function is_inline_checker(@nospecialize stmt)
+            isa(stmt, GlobalRef) && stmt.name === :inline_checker
+        end
+    end
+
+    @testset "@nospecialize" begin
+        @eval m begin
+            function nospecialize(@nospecialize a)
+                c = isa(a, Function)
+                inline_checker(c)
+            end
+
+            @inline function inline_nospecialize(@nospecialize a)
+                c = isa(a, Function)
+                inline_checker(c)
+            end
+        end
+
+        # `@nospecialize` should suppress runtime dispatches of `nospecialize`
+        m.invokef(m.nospecialize, dispatchonly)
```

Comment:
To prevent possible confusion with `Method.nospecialize`, should you use a capital `M` for modules?

Reply:
yeah, that sounds better